titan-compiler/coder.lua

```diff
@@ -230,14 +510,227 @@ generate_exp = function(exp) -- TODO
     elseif tag == ast.Exp.Initlist then
         error("not implemented yet")
 
-    elseif tag == ast.Exp.Unop then
+    elseif tag == ast.Exp.Call then
         error("not implemented yet")
 
+    elseif tag == ast.Exp.Var then
+        return generate_var(exp.var)
+
+    elseif tag == ast.Exp.Unop then
+        local cstats, cvalue = generate_exp(exp.exp)
+
+        local op = exp.op
+        if op == "#" then
+            error("not implemented yet")
+
+        elseif op == "-" then
+            return cstats, "(".."-"..cvalue..")"
+
+        elseif op == "~" then
+            return cstats, "(".."~"..cvalue..")"
+
+        elseif op == "not" then
+            return cstats, "(".."!"..cvalue..")"
+
+        else
+            error("impossible")
+        end
+
     elseif tag == ast.Exp.Concat then
         error("not implemented yet")
 
     elseif tag == ast.Exp.Binop then
-        error("not implemented yet")
+        local lhs_cstats, lhs_cvalue = generate_exp(exp.lhs)
+        local rhs_cstats, rhs_cvalue = generate_exp(exp.rhs)
+
+        -- Lua's arithmetic and bitwise operations for integers happen with
+        -- unsigned integers, to ensure 2's compliment behavior and avoid
+        -- undefined behavior.
+        local function intop(op)
+            local cstats = lhs_cstats..rhs_cstats
+            local cvalue = util.render("intop(${OP}, ${LHS}, ${RHS})", {
+                OP=op, LHS=lhs_cvalue, RHS=rhs_cvalue })
+            return cstats, cvalue
+        end
+
+        -- Relational operators, and basic float operations don't convert their
+        -- parameters
+        local function binop(op)
+            local cstats = lhs_cstats..rhs_cstats
+            local cvalue = util.render("((${LHS})${OP}(${RHS}))", {
+                OP=op, LHS=lhs_cvalue, RHS=rhs_cvalue })
+            return cstats, cvalue
+        end
+
+        local ltyp = exp.lhs._type._tag
+        local rtyp = exp.rhs._type._tag
+
+        local op = exp.op
+        if     op == "+" then
+            if     ltyp == types.T.Integer and rtyp == types.T.Integer then
```

Comment:
There is a lot of repetition here. I think we should create an abstraction (but this could be in another PR).